milestones/api.py

```diff
@@ -0,0 +1,312 @@
+    milestone = {
+        'namespace': namespace,
+    }
+    _validate_milestone(milestone)
+    return data.fetch_milestones(milestone)
+
+
+def remove_milestone(milestone_id):
+    """
+    Removes the specified milestone
+    """
+    milestone = {
+        'id': milestone_id,
+    }
+    data.delete_milestone(milestone)
+
+
+def add_course_milestone(course_key, relationship, milestone):
+    """
+    Adds a course-milestone link to the system
+    'relationship': string value (eg: 'requires')
+    """
+    _validate_course_key(course_key)
+    _validate_milestone_relationship_type(relationship)
+    _validate_milestone(milestone)
+    data.create_course_milestone(course_key=course_key, relationship=relationship, milestone=milestone)
+
+
+def get_course_milestones(course_key, relationship=None):
+    """
+    Retrieves the set of milestones for a given course
+    'relationship': optional filter on milestone relationship type (string, eg: 'requires')
+    Returns an array of dicts containing milestones
+    """
+    _validate_course_key(course_key)
+
+    if relationship is not None:
+        _validate_milestone_relationship_type(relationship)
+    return data.fetch_courses_milestones(course_keys=[course_key], relationship=relationship)
+
+
+def get_course_required_milestones(course_key, user):
+    """
+    Retrieves the set of required milestones for a given course that a user has not yet collected
+    """
+    _validate_course_key(course_key)
+    _validate_user(user)
+    required_milestones = data.fetch_courses_milestones([course_key], 'requires', user)
+    return required_milestones
+
+
+def get_course_milestones_fulfillment_paths(course_key, user):
+    """
+    Returns a collection composed of the possible fulfillment/collection opportunites
+    """
+    _validate_course_key(course_key)
+    _validate_user(user)
+
+    # Retrieve the outstanding milestones for this course, for this user
+    required_milestones = data.fetch_courses_milestones([course_key], 'requires', user)
+
+    # Build the set of fulfillment paths for the outstanding milestones
+    fulfillment_paths = {}
+    for milestone in required_milestones:
+        dict_key = 'milestone_{}'.format(milestone['id'])
+        fulfillment_paths[dict_key] = {}
+        milestone_courses = data.fetch_milestone_courses(milestone, 'fulfills')
+        if len(milestone_courses):
+            fulfillment_paths[dict_key]['courses'] = milestone_courses
+        milestone_course_content = data.fetch_milestone_course_content(milestone, 'fulfills')
+        if len(milestone_course_content):
+            fulfillment_paths[dict_key]['content'] = milestone_course_content
+    return fulfillment_paths
+
+
+def get_courses_milestones(course_keys, relationship=None, user=None):
+    """
+    Retrieves the set of milestones for list of courses
+    'relationship': optional filter on milestone relationship type (string, eg: 'requires')
+    Returns an array of dicts containing milestones
```

Comment:
you might want to add a comment about how the optional 'user' parameter is used - I presume it is a filter, but it might be good to have some semantic definitions around it.